### ISSUE TITLE
Connection workflow, MVC style

### ISSUE THREAD

**Issue by cadilhac** [JoeMayo/LinqToTwitter#49]
Hi,

The first time a user wants to authenticate, it works well. Then, if he logs ou of my app and tries a second time to sign in with twitter I get an exception here:

`await auth.BeginAuthorizationAsync(new Uri(twitterCallbackUrl))`

> Value cannot be null or empty.
> Parameter name: url
> in RedirectResult

Of course, twitterCallbackUrl did not change and it's the same as the first time.
If I recompile and run my app, then it works again once.

Note also that even if my app has been allowed in his account, the twitter form to allow the app appears again on subsequent logins. On facebook, after the initial permission, the facebook form does not appear again and it goes directly to the redirect url.

**Comment by cadilhac** [comment]
About my last note, it seems MvcSignInAuthorizer is the way to go. Am I right?
If yes, the only problem left is the exception.

**Comment by JoeMayo** [comment]
Hi,

Here's Twitter's OAuth page, where you can review and get an idea of which authorization type works best for your particular application:

https://dev.twitter.com/oauth

A typical reason for login issues with ASP.NET applications is due to Session state. If you're using SessionStateCredentialStore, debug and check to see if your credentials are present the second time around. They might not be there if you're using the default inMemory session state, which is cleared out if your Web server recycles. This can be fixed by switching to SQLServer or StateServer, which you can set via the mode attribute of the sessionState element in your Web.config file.

[@JoeMayo](https://twitter.com/JoeMayo)

**Comment by cadilhac** [comment]
ok, I checked into your code and I see that AspNetAuthorizer.BeginAuthorizeAsync will simply return without calling GoToTwitterAuthorization and that's why RedirectResult(_authUrl) throws a null ref exception. So I should check if the credentials are still valid or not before calling your code, or reset them.

But are you sure this is the way to handle this case? Should you not throw a proper exception like "Already authenticated" instead?

**Comment by JoeMayo** [comment]
Thanks for explaining that - it's easy to get so close to the code that I don't see all scenarios. :) I'll take a look at it.

**Comment by jlaustill** [comment]
I think I'm running into this as well.  I have had a service running for MONTHS without issue, then it just started failing with this exact same error.  Recompiling doesn't seem to fix it for me though, so I'm going to be spending the day trying to figure it out.

**Comment by jlaustill** [comment]
Ok, I tracked this down on my end, it wasn't an issue with linqtotwitter at all, it was the fault of my marketing team changing our username without letting me know.  It would have been nice if the error told me that my username wasn't found instead of just a null ref exception.  But I got it figured out.  I was also having an issue after updating to the latest linqtotwitter version(did so as a debugging step to see if maybe something had just changed in the twitter api that you updated for) that I had to target my project to .net 4.6.1 instead of 4.5.1, which isn't a huge deal either.  So I don't know if this contributes to the discussion at all, but there you have it.  Thanks for your hard work on this library!

**Comment by JoeMayo** [comment]
Hi jlaustill,

Thanks for the feedback.  I'll look at that UserName issue too.

[@JoeMayo](https://twitter.com/JoeMayo)

**Comment by jlaustill** [comment]
@JoeMayo 

Here's the query I was running in case it helps.

```
from user in twitterCtx.User
                  where user.Type == UserType.Lookup && user.ScreenNameList == "username"
                  select user
```

**Comment by AndyCJ** [comment]
Hi, we're seeing this issue too.

In our application a user can "link" (login via oauth and we capture the access tokens) different twitter accounts to our app. The problem arises when a user tries to link a twitter account twice within the same session. We're using the SessionStateCredentialStore with the MvcAuthorizer.


It would appear this line is the problem: 
`if (CredentialStore.HasAllCredentials()) return;`
https://github.com/JoeMayo/LinqToTwitter/blob/master/New/Libraries/LinqToTwitter.AspNet/AspNetAuthorizer.cs#L69


That means this line is skipped:
`string authUrl = PrepareAuthorizeUrl(ForceLogin);`
https://github.com/JoeMayo/LinqToTwitter/blob/master/New/Libraries/LinqToTwitter.AspNet/AspNetAuthorizer.cs#L80


So when this line is executed, the _authUrl is empty:
`return new RedirectResult(_authUrl);`
https://github.com/JoeMayo/LinqToTwitter/blob/master/New/Libraries/LinqToTwitter.AspNet/MvcAuthorizer.cs#L25


I think the solution is as cadilhac suggested above, which is to throw an exception, rather than simply return. i.e.
`if (CredentialStore.HasAllCredentials()) throw new ReliventException("Already Authorised.");`


The example code in the documentation would then need updating to something along the lines of:
```
var authorizer = new MvcAuthorizer
{
	CredentialStore = new SessionStateCredentialStore
	{
		ConsumerKey = consumerKey,
		ConsumerSecret = consumerSecret
	}
};

if (authorizer.CredentialStore.HasAllCredentials())
{
	//Use the existing credentials
	
	return null;// return something relivent
}
else
{
	var twitterCallbackUrl = Request
		.Url
		.ToString()
		.Replace("Begin", "Complete");

	return await authorizer.BeginAuthorizationAsync(new Uri(twitterCallbackUrl));
}
```

Hope this helps!

**Comment by AndyCJ** [comment]
I recently revisited this issue and realised what I wrote doesn't work for my situation. When the user wants to log in for the second time it might be for a different account, so it's not enough to simply reuse the credentials already provided.

I realised that I needed to wipe out the existing credentials on logout so that the authorization process could run again from scratch.

I initially tried to use ClearAsync but this is not implemented in the InMemoryCredentialStore, nor is it overridden in the SessionStateCredentialStore. https://github.com/JoeMayo/LinqToTwitter/blob/master/New/Libraries/LinqToTwitterPcl/Security/InMemoryCredentialStore.cs#L47

So instead I wrote a quick method to wipe the information:

        private void ClearSessionStateCredentialStore(SessionStateCredentialStore sessionStateCredentialStore)
        {
            sessionStateCredentialStore.OAuthToken = string.Empty;
            sessionStateCredentialStore.OAuthTokenSecret = string.Empty;
            sessionStateCredentialStore.ScreenName = string.Empty;
            sessionStateCredentialStore.UserID = 0ul;
        }

So if the user then tries to log in again `if (CredentialStore.HasAllCredentials()) return;` is false, and the authorization process is allowed to continue.

**Comment by sohaibjaved44** [comment]
How I can get access token when it reaches call back url

**Comment by soongsta** [comment]
Adding to what AndyCJ, this worked for me
                    CredentialStore = new SessionStateCredentialStore()
                    {
                        OAuthToken = string.Empty,
                        OAuthTokenSecret = string.Empty,
                        ScreenName = string.Empty,
                        UserID = 0ul,
                        ConsumerKey = xxxxx,
                        ConsumerSecret = yyy
                    }

**Comment by WarrenMT** [comment]
@AndyCJ Thank you very much. Adding that code saved me a lot of pain and time. Super Upvote

**Comment by JoeMayo** [comment]
There will be two changes in the next release. The first throws an exception when someone tries to authorize and they already have all four credentials. There's an extensive description in the InvalidOperationException. The second is that I implemented ClearAsync in all ICredentialStore types to simplify clearing old credentials. I'm working through remaining issues and plan to release soon.

**Comment by JoeMayo** [comment]
This was fixed in [LINQ to Twitter v4.2.0](https://github.com/JoeMayo/LinqToTwitter/releases/tag/v4.2.0).

[@JoeMayo](https://twitter.com/JoeMayo)
